### PR TITLE
[#4485] Adjusting the address isolation logic of the servicecomb engine registration and config center

### PR DESCRIPTION
**Pull request by chengyouling**

### CODE REVIEW

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java [hunk 1]
```diff
@@ -53,42 +51,37 @@ public class AbstractAddressManager {
...
+
+  private final List<String> default_isolation_address = new ArrayList<>();
 
   private int index;
 
   private String projectName;
 
   // if address in same zone will be true; others will be false.
-  private final Map<String, Boolean> addressCategory = new HashMap<>();
+  private final Set<String> addressCategory = new HashSet<>();
```
Comment (liubao68) [inline clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java:67]: code comments not correct.

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java [hunk 2]
```diff
@@ -53,42 +51,37 @@ public class AbstractAddressManager {
...
 
-  private static final int DEFAULT_METRICS_WINDOW_TIME = 1;
+  private static final int DEFAULT_ADDRESS_CHECK_TIME = 30;
 
   private static final int ISOLATION_THRESHOLD = 3;
 
-  private List<String> addresses = new ArrayList<>();
+  private volatile List<String> addresses = new ArrayList<>();
+
+  private final List<String> default_isolation_address = new ArrayList<>();
```
Comment (liubao68) [inline clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java:60]: Naming convention

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java [hunk 3]
```diff
@@ -98,25 +91,33 @@ public class AbstractAddressManager {
   public AbstractAddressManager(List<String> addresses) {
     this.projectName = DEFAULT_PROJECT;
     this.addresses.addAll(addresses);
-    this.defaultAddress.addAll(addresses);
-    this.index = addresses.size() > 0 ? random.nextInt(addresses.size()) : 0;
+    this.addressCategory.addAll(addresses);
+    this.index = !addresses.isEmpty() ? random.nextInt(addresses.size()) : 0;
```
[the remark this answers is not in the document]
Comment (chengyouling) [reply clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java:95]: only default scene use addressCategory for address choose in address() method, and in this scence, addresses and addressCategory may have same size

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java [hunk 4]
```diff
@@ -175,27 +161,34 @@ protected String formatAddress(String address) {
...
   }
 
-  private String getAvailableZoneAddress() {
-    List<String> addresses = getAvailableZoneIpPorts();
+  private String getDefaultAddress() {
     if (!addresses.isEmpty()) {
       return getCurrentAddress(addresses);
     }
-    return getInitAddress();
+    LOGGER.warn("all address is isolation, please check server status.");
```
Comment (liubao68) [inline clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java:180]: All addresses are isolated.

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java [hunk 5]
```diff
@@ -175,27 +161,34 @@ protected String formatAddress(String address) {
...
-  // when all available address is fail, it will use all the initial addresses for polling.
-  private String getInitAddress() {
-    if (addresses.isEmpty()) {
-      return null;
+  private String getAvailableZoneAddress() {
+    List<String> zoneOrRegionAddress = getZoneOrRegionAddress();
+    if (!zoneOrRegionAddress.isEmpty()) {
+      return getCurrentAddress(zoneOrRegionAddress);
     }
+    LOGGER.warn("all auto discovery address is isolation, please check server status.");
```
Comment (liubao68) [inline clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java:190]: Refer to above

Comment (chengyouling) [reply clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java]: fixed

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java (hunk 4 above)
Comment (chengyouling) [reply clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java]: fixed

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java (hunk 1 above)
Comment (chengyouling) [reply clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java]: fixed

File: clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java (hunk 2 above)
Comment (chengyouling) [reply clients/http-client-common/src/main/java/org/apache/servicecomb/http/client/common/AbstractAddressManager.java]: fixed

**Pull request merged**